File: java/graphs/flows/MaxFlowDinic.java
package graphs.flows;

import java.util.*;
import java.util.stream.Stream;

// https://en.wikipedia.org/wiki/Dinic%27s_algorithm in O(V^2 * E)
public class MaxFlowDinic {
    List<Edge>[] graph;
    int[] dist;

    public MaxFlowDinic(int nodes) {
        graph = Stream.generate(ArrayList::new).limit(nodes).toArray(List[] ::new);
        dist = new int[nodes];
    }

    class Edge {
        int t, rev, cap, f;

        public Edge(int t, int rev, int cap) {
            this.t = t;
            this.rev = rev;
            this.cap = cap;
        }
    }

    public void addBidiEdge(int s, int t, int cap) {
        graph[s].add(new Edge(t, graph[t].size(), cap));
        graph[t].add(new Edge(s, graph[s].size() - 1, 0));
    }

    boolean dinicBfs(int src, int dest) {
        Arrays.fill(dist, -1);
        dist[src] = 0;
        int[] q = new int[graph.length];
        int sizeQ = 0;
        q[sizeQ++] = src;
        for (int i = 0; i < sizeQ; i++) {
            int u = q[i];
            for (Edge e : graph[u]) {
                if (dist[e.t] < 0 && e.f < e.cap) {
                    dist[e.t] = dist[u] + 1;
                    q[sizeQ++] = e.t;
                }
            }
        }
        return dist[dest] >= 0;
    }

    int dinicDfs(int[] ptr, int dest, int u, int f) {
        if (u == dest)
            return f;
        for (; ptr[u] < graph[u].size(); ++ptr[u]) {
            Edge e = graph[u].get(ptr[u]);
            if (dist[e.t] == dist[u] + 1 && e.f < e.cap) {
                int df = dinicDfs(ptr, dest, e.t, Math.min(f, e.cap - e.f));
                if (df > 0) {
                    e.f += df;
                    graph[e.t].get(e.rev).f -= df;
                    return df;
                }
            }
        }
        return 0;
    }

    public int maxFlow(int src, int dest) {
        int flow = 0;
        while (dinicBfs(src, dest)) {
            int[] ptr = new int[graph.length];
            for (int df; (df = dinicDfs(ptr, dest, src, Integer.MAX_VALUE)) != 0; flow += df)
                ;
        }
        return flow;
    }

    // invoke after maxFlow()
    public boolean[] minCut() {
        boolean[] cut = new boolean[graph.length];
        for (int i = 0; i < cut.length; ++i) cut[i] = dist[i] != -1;
        return cut;
    }

    void clearFlow() {
        for (List<Edge> edges : graph)
            for (Edge edge : edges) edge.f = 0;
    }

    // Usage example
    public static void main(String[] args) {
        MaxFlowDinic flow = new MaxFlowDinic(3);
        flow.addBidiEdge(0, 1, 3);
        flow.addBidiEdge(0, 2, 2);
        flow.addBidiEdge(1, 2, 2);
        System.out.println(4 == flow.maxFlow(0, 2));
    }
}

File: golang/graphs/flows/max_flow_dinic.go
package flows

import "math"

// MaxFlowDinic implements Dinic's algorithm for maximum flow in O(V^2 * E)
// https://en.wikipedia.org/wiki/Dinic%27s_algorithm
type MaxFlowDinic struct {
	graph [][]Edge
	dist  []int
}

type Edge struct {
	to, rev, cap, f int
}

func NewMaxFlowDinic(n int) *MaxFlowDinic {
	return &MaxFlowDinic{
		graph: make([][]Edge, n),
		dist:  make([]int, n),
	}
}

func (mf *MaxFlowDinic) AddEdge(s, t, cap int) {
	mf.graph[s] = append(mf.graph[s], Edge{t, len(mf.graph[t]), cap, 0})
	mf.graph[t] = append(mf.graph[t], Edge{s, len(mf.graph[s]) - 1, 0, 0})
}

func (mf *MaxFlowDinic) bfs(src, dest int) bool {
	for i := range mf.dist {
		mf.dist[i] = -1
	}
	mf.dist[src] = 0
	queue := []int{src}
	for i := 0; i < len(queue); i++ {
		u := queue[i]
		for _, e := range mf.graph[u] {
			if mf.dist[e.to] < 0 && e.f < e.cap {
				mf.dist[e.to] = mf.dist[u] + 1
				queue = append(queue, e.to)
			}
		}
	}
	return mf.dist[dest] >= 0
}

func (mf *MaxFlowDinic) dfs(ptr []int, dest, u, f int) int {
	if u == dest {
		return f
	}
	for ; ptr[u] < len(mf.graph[u]); ptr[u]++ {
		e := &mf.graph[u][ptr[u]]
		if mf.dist[e.to] == mf.dist[u]+1 && e.f < e.cap {
			df := mf.dfs(ptr, dest, e.to, min(f, e.cap-e.f))
			if df > 0 {
				e.f += df
				mf.graph[e.to][e.rev].f -= df
				return df
			}
		}
	}
	return 0
}

func (mf *MaxFlowDinic) MaxFlow(src, dest int) int {
	flow := 0
	for mf.bfs(src, dest) {
		ptr := make([]int, len(mf.graph))
		for {
			df := mf.dfs(ptr, dest, src, math.MaxInt32)
			if df == 0 {
				break
			}
			flow += df
		}
	}
	return flow
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
